fitlins/base.py

```diff
@@ -0,0 +1,203 @@
+        # names = [col for col in confounds.columns
+        #          if col.startswith('NonSteadyStateOutlier') or
+        #          col in block.model['variables']]
+        img = nb.load(fname)
+        TR = img.header.get_zooms()[3]
+        vols = img.shape[3]
+
+        mat = dm.make_design_matrix(np.arange(vols) * TR,
+                                    paradigm.rename(columns={'condition': 'trial_type'}),
+                                    drift_model=None,
+                                    # add_regs=confounds[names],
+                                    # add_reg_names=names,
+                                    )
+
+        out_dir = deriv_dir
+        if 'subject' in ents:
+            out_dir = op.join(out_dir, 'sub-' + ents['subject'])
+        if 'session' in ents:
+            out_dir = op.join(out_dir, 'ses-' + ents['session'])
+
+        os.makedirs(out_dir, exist_ok=True)
+
+        base = op.basename(fname)
+        design_fname = op.join(out_dir, base.replace('_preproc.nii.gz', '_design.tsv'))
+
+        mat.to_csv(design_fname, sep='\t')
+
+        brainmask = analysis.layout.get(type='brainmask', **ents)[0]
+        fmri_glm = None
+
+        for contrast in block.contrasts:
+            stat_fname = op.join(out_dir,
+                                 base.replace('_preproc.nii.gz',
+                                              '_contrast-{}_stat.nii.gz'.format(
+                                                  snake_to_camel(contrast['name']))))
+            out_imgs.setdefault(contrast['name'], []).append(stat_fname)
+
+            if op.exists(stat_fname):
+                continue
+
+            if fmri_glm is None:
+                fmri_glm = level1.FirstLevelModel(mask=brainmask.filename)
+                fmri_glm.fit(fname, design_matrices=mat)
+
+            indices = [mat.columns.get_loc(cond)
+                       for cond in contrast['condition_list']]
+
+            weights = np.zeros(len(mat.columns))
+            weights[indices] = contrast['weights']
+
+            stat = fmri_glm.compute_contrast(weights, {'T': 't', 'F': 'F'}[contrast['type']])
+            stat.to_filename(stat_fname)
+
+    return out_imgs
+
+
+def second_level(analysis, block, in_imgs, deriv_dir):
+    out_imgs = {}
+
+    for xform in block.transformations:
+        if xform['name'] == 'split':
+            for in_col in xform['input']:
+                img_set = in_imgs[in_col]
+                by = xform['by']
+                fmt = '{}-{{}}'.format('ses' if by == 'session' else 'sub').format
+                splitter = {'session': analysis.layout.get_sessions,
+                            'subject': analysis.layout.get_subjects}[by]()
+                for var in splitter:
+                    out_imgs['{}.{}'.format(var, in_col)] = [
+                        img for img in in_imgs[in_col]
+                        if fmt(var) in img]
```

Comment:
The above handles splitting. Given a dictionary of filenames of statistical maps, I can split out by session or subject. Does this make sense, or is there something internal to the pybids code that I can piggy-back off of? And I haven't decided how I'll group by subject to match the `.get_Xy()` blocks. I'm trying not to code too closely to the sample model.

Reply:
Not sure I follow... can you give an example of where you'd need to do this?

Reply:
In general, images shouldn't need to be transformed in any way as far as I can see, aside from entering into contrasts. In principle we *could* support applying transformations to images, but that seems like a pretty uncommon use case, and my inclination would be to defer any work on that for quite a while...

Reply:
I don't mean transforming the images directly, but to the extent that transforms describe ways of splitting and selecting lower-level maps to perform higher-level contrasts on, I need to be able to apply them. For instance, I have a series of images tagged with the contrast `finger_vs_others`, and now two levels later I take a contrast of `test.finger_vs_others` and `retest.finger_vs_others`. I need to associate each pair of maps correctly, and there should (in this case) be a pair from each subject.

Reply:
I'm still not sure you need to do anything more than apply the contrasts. Consider your `finger_vs_others` example. Let's take the case of a single subject, where we're trying to apply the session-level `split` transformation that takes `finger_vs_others` as input, and splits on `session`. Initially, we have two input rows (the `finger_vs_others` image from each of the sessions) and a single column (`'finger_vs_others'`). The design matrix (plus other information we retain internally, namely session) is just:

|image_id|session|finger_vs_others|
|--|--|--|
|1|'test'|1|
|2|'retest'|1|

When we apply the split transformation, we're basically just taking the product of (dummy-coded) `session` and `finger_vs_others`. This gives us the following design matrix:

|image_id|test_finger_vs_others|retest_finger_vs_others|
|--|--|--|
|1|1|0|
|2|0|1|

When you apply the contrasts for these two columns, you automatically get two new images with the appropriate names, and the contents should automatically be correct also, because only the first image is used in the first column, and only the second image is used in the second column.

Does that work?

Reply:
So it sounds like you're proposing performing a null (or identity?) contrast on levels that don't otherwise have contrasts. Should `.get_contrasts()` then provide that?

Reply:
It should, yes. I thought I had already implemented that, and that it was the default, but maybe not. Will take a look.

Reply:
Just checked... it *should* be creating contrasts for all design matrix columns by default (there's a `keep_input_columns` argument that's `True` by default). If that's not working, it's a bug.